packages/core/src/DataSchema.ts

```diff
@@ -0,0 +1,118 @@
+    }
+
+    parseSchema(schema: any[]): any[] {
+        let sumOffset = 0
+        let replOffset = 0
+        const maxSumOffset = 253 * this.config.SUM_FIELD_COUNT
+        const maxReplOffset =
+            253 * (this.config.FIELD_COUNT - this.config.SUM_FIELD_COUNT)
+
+        return schema.map((field, idx) => {
+            const { name, type, updateBy, ...extranousFields } = field
+            const match = type.match(/^uint(\d+)$/)
+            if (match === null)
+                throw new Error(`Invalid type for field ${name}: "${type}"`)
+            if (Object.keys(extranousFields).length !== 0) {
+                throw new Error(
+                    `Invalid fields included for field ${name}: [${Object.keys(
+                        extranousFields
+                    )}]`
+                )
+            }
+
+            const duplicateEntry = schema.find(
+                (x, i) => x.name == field.name && idx != i
+            )
+            if (duplicateEntry) {
+                throw Error(
+                    `Schema includes a duplicate entry: "${duplicateEntry.name}"`
+                )
+            }
+
+            const bits = +match[1]
+            if (bits < 1 || bits > 253)
+                throw new Error(`Invalid uint size for field ${name}: ${bits}`)
+            if (updateBy === 'sum') {
+                if (
+                    Math.floor(sumOffset / 253) !==
+                    Math.floor((sumOffset + bits - 1) / 253)
+                ) {
+                    sumOffset += 253 - (sumOffset % 253)
+                }
+                if (sumOffset + bits > maxSumOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+                const dataIndex = Math.floor(sumOffset / 253)
+                const offset = sumOffset % 253
+                sumOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            } else if (updateBy === 'replace') {
+                if (bits !== 253 - this.config.REPL_NONCE_BITS)
+                    throw new Error(
+                        `Invalid uint size for field ${name}: ${bits}`
+                    )
+                if (replOffset + bits > maxReplOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+
+                const dataIndex =
+                    this.config.SUM_FIELD_COUNT + Math.floor(replOffset / 253)
+                const offset = replOffset % 253
+                replOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            }
+            throw new Error(
+                `Invalid updateBy strategy for field ${name}: "${updateBy}"`
+            )
+        })
+    }
+
+    update(name: string, v: bigint): bigint {
```

Comment:
Instead of having an `update` function to transform the data, we should have a function `buildAttestations` to apply some changes to a given data.

This is complex and we'll probably want to adjust the contracts to help with this (e.g. a multi-attest function to change multiple fields in 1 call).

So this could be something like

```ts
type Attestation = { fieldIndex: number, change: bigint }

buildAttestation(change: { name: string, v: bigint }): Attestation
buildAttestations(changes: { name: string, v: bigint }[]): Attestation[]
```

Keep in mind that if there are multiple changes to a single field they can be combined into a single attestation.

Reply:
This is also why I don't want to track `data` as a property on this class. The attester isn't meant to know the user data, and will often attest without knowing the full user data. The attester will also receive partial proofs of data. e.g. a user might prove bits 32-64 of field index 2 to the attester and leave all other fields as 0. There's not much use in storing that on the instance of the class. It's better to think of this class like the [ethers abi coder](https://docs.ethers.org/v6/api/abi/abi-coder/).

Reply:
Done, `buildAttestation` and `buildAttestations` now included. `buildAttestation` combines taking all changes that correspond to the same `fieldIndex` and altering `change` value of the attestation going to that field. Additionally, multiple `sum` attestations are combined, where the additions are performed in `buildAttestations` with error handling. Both `buildAttestation` and `buildAttestations` return an `Attestation {fieldIndex, change}` type struct. `fieldIndex` is found through a lookup of the data schema created in `parseSchema`